### PR TITLE
Ticket1355 show iocs from components

### PR DESCRIPTION
**Pull request by ThomasLohnert**
### Description of work

When adding a component to a configuration, IOCs in said component are added to the configuration with all their settings, macros etc (same as blocks).

System test: https://github.com/ISISComputingGroup/ibex_system_tests/pull/44

### To test

https://github.com/ISISComputingGroup/IBEX/issues/1355

### Acceptance criteria

* Component IOCs show up in IOC tab after adding component to config.
* Settings of component IOC can be viewed, but not edited
* IOC is started with settings as set in component.

---

#### Code Review

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/GUI-Coding-Conventions)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit tests in place? Are the unit tests small and test the a class in isolation?
- [x] Are there automated [system tests](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/System-Testing-with-RCPTT) in place? Do they test a minimal set of functionality and leave the gui as close as possible to its original state?
- [ ] Has the [manual system tests spreadsheet](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Manual-system-tests) been updated?
    - Manual system tests for the functionality should be added if there are no automated tests
    - Manual system tests can be removed from the template if they are covered by suitable automated tests
- [ ] Did any existing system test break as a result of the current changes? 
- [x] Have the changes been documented in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev). If so, do they describe the changes appropriately?
- [x] If an OPI has been modified, does it conform to the [style guidelines](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/OPI-Creation)? There is a script called `check_opi_format.py` to help with this.

### Functional Tests

- [x] Do changes function as described? Add comments below that describe the tests performed.
- [x] How do the changes handle unexpected situations, e.g. bad input?
- [x] Has developer documentation been updated if required?

### Final Steps
- [x] Reviewer has moved the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev) entry for this ticket in the "Changes merged into master" section

### CODE REVIEW

File: base/uk.ac.stfc.isis.ibex.configserver/src/uk/ac/stfc/isis/ibex/configserver/editing/EditableConfiguration.java [hunk 1]
```diff
@@ -298,7 +329,7 @@ public Group apply(EditableGroup group) {
      * @return The IOCs associated with the configuration
      */
     private Collection<Ioc> getIocs() {
```
Comment (DominicOram) [inline base/uk.ac.stfc.isis.ibex.configserver/src/uk/ac/stfc/isis/ibex/configserver/editing/EditableConfiguration.java]: Looking through this code, I don't see where its used that getSelectedIocs couldn't be instead. Can we remove it?

File: base/uk.ac.stfc.isis.ibex.configserver/src/uk/ac/stfc/isis/ibex/configserver/editing/EditableConfiguration.java [hunk 2]
```diff
@@ -381,17 +412,20 @@ public void removeIocs(List<EditableIoc> iocs) {
      * @return The editable IOCs associated with the configuration
```
Comment (DominicOram) [inline base/uk.ac.stfc.isis.ibex.configserver/src/uk/ac/stfc/isis/ibex/configserver/editing/EditableConfiguration.java]: I think this comment is now incorrect, we're returning all IOCs, editable or not

File: base/uk.ac.stfc.isis.ibex.configserver/src/uk/ac/stfc/isis/ibex/configserver/editing/EditableConfiguration.java [hunk 3]
```diff
@@ -381,17 +412,20 @@ public void removeIocs(List<EditableIoc> iocs) {
...
      */
     public Collection<EditableIoc> getSelectedIocs() {
```
Comment (DominicOram) [inline base/uk.ac.stfc.isis.ibex.configserver/src/uk/ac/stfc/isis/ibex/configserver/editing/EditableConfiguration.java]: Can you rename this function?

File: base/uk.ac.stfc.isis.ibex.ui.configserver/src/uk/ac/stfc/isis/ibex/ui/configserver/editing/iocs/dialog/IocDialog.java [hunk 4]
```diff
@@ -114,87 +64,47 @@ private void nextPage() {
...
      * @param ioc
      *            The IOC to add.
-     * @param isNew
-     *            Flag indicating whether IOC to edit is newly added or
-     *            pre-existing.
+     * @param readOnly
+     *            Specifies whether the settings are editable
      */
-    public IocDialog(Shell parent, EditableConfiguration config, EditableIoc ioc, boolean isNew) {
+    public IocDialog(Shell parent, EditableConfiguration config, EditableIoc ioc, boolean readOnly) {
```
Comment (DominicOram) [inline base/uk.ac.stfc.isis.ibex.ui.configserver/src/uk/ac/stfc/isis/ibex/ui/configserver/editing/iocs/dialog/IocDialog.java]: Can we use ioc.isEditable() instead? I think that means we could remove readOnly from here and in IocOverviewPanel

File: base/uk.ac.stfc.isis.ibex.ui.configserver/src/uk/ac/stfc/isis/ibex/ui/configserver/editing/iocs/dialog/AddIocDialog.java [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+            @Override
+            public void propertyChange(PropertyChangeEvent evt) {
+                nextPage();
+            }
+        });
+    }
+
+    @Override
+    protected Control createDialogArea(Composite parent) {
```
Comment (DominicOram) [inline base/uk.ac.stfc.isis.ibex.ui.configserver/src/uk/ac/stfc/isis/ibex/ui/configserver/editing/iocs/dialog/AddIocDialog.java]: This seems to share quite a bit of code with the same method in the parent, can we reduce the duplication?

**Comment by DominicOram** [comment]
All functionality is good, a few minor issues, both here and in the system tests. I've commented on them in the code.

File: base/uk.ac.stfc.isis.ibex.configserver/src/uk/ac/stfc/isis/ibex/configserver/editing/EditableConfiguration.java (hunk 1 above)
Comment (ThomasLohnert) [inline base/uk.ac.stfc.isis.ibex.configserver/src/uk/ac/stfc/isis/ibex/configserver/editing/EditableConfiguration.java]: This method is used when transforming an EditableConfiguration to a Configuration (the format in which it is then sent to the blockserver), to populate that new Configuration object with the right types of elements - same as getBlocks, getGroups getComponents in this class. I can rename those methods to transformIocs etc. to more accurately reflect the purpose of the methods. Does that sound sensible?

**Pull request merged**